flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -99,87 +75,360 @@ class DoConcurrentConversion : public mlir::OpConversionPattern<fir::DoLoopOp> {
-    //                   we properly implement and test it for more complicated
-    //                   `do concurrent` loops.
-    //   1. **Stage 2**: Using the collected operations, create and populate an
-    //                   `omp.private {type=private}` op to server as the
-    //                   delayed privatizer for the new work-sharing loop.
-
-    // For the induction variable, we need to privative its allocation and
-    // binding inside the parallel region.
-    llvm::SmallSetVector<mlir::Operation *, 2> workList;
-    // Therefore, we first discover the induction variable by discovering
-    // `fir.store`s where the source is the loop's block argument.
-    workList.insert(doLoop.getInductionVar().getUsers().begin(),
-                    doLoop.getInductionVar().getUsers().end());
-    llvm::SmallSetVector<fir::StoreOp, 2> inductionVarTargetStores;
-
-    // Walk the def-chain of the loop's block argument until we hit `fir.store`.
-    while (!workList.empty()) {
-      mlir::Operation *item = workList.front();
-
-      if (auto storeOp = mlir::dyn_cast<fir::StoreOp>(item)) {
-        inductionVarTargetStores.insert(storeOp);
-      } else {
-        workList.insert(item->getUsers().begin(), item->getUsers().end());
-      }
+    if (mapToDevice) {
+      mlir::omp::TargetClauseOps clauseOps;
+      for (mlir::Value liveIn : liveIns)
+        clauseOps.mapVars.push_back(genMapInfoOpForLiveIn(rewriter, liveIn));
+
+      targetOp =
+          genTargetOp(doLoop.getLoc(), rewriter, mapper, liveIns, clauseOps);
+      genTeamsOp(doLoop.getLoc(), rewriter);
+      genDistributeOp(doLoop.getLoc(), rewriter);
+    }
+
+    genParallelOp(doLoop.getLoc(), rewriter, liveIns, mapper);
+    genWsLoopOp(rewriter, doLoop, mapper);
+
+    // Now that we created the nested `ws.loop` op, we can the `target` op's
+    // trip count.
+    if (mapToDevice) {
+      rewriter.setInsertionPoint(targetOp);
+      mlir::ModuleOp parentModule = doLoop->getParentOfType<mlir::ModuleOp>();
+      fir::FirOpBuilder firBuilder(rewriter, fir::getKindMapping(parentModule));
+
+      mlir::omp::CollapseClauseOps collapseClauseOps;
+      collapseClauseOps.loopLBVar.push_back(lbOp->getResult(0));
+      collapseClauseOps.loopUBVar.push_back(ubOp->getResult(0));
+      collapseClauseOps.loopStepVar.push_back(stepOp->getResult(0));
+
+      mlir::cast<mlir::omp::TargetOp>(targetOp).getTripCountMutable().assign(
+          Fortran::lower::omp::calculateTripCount(firBuilder, doLoop.getLoc(),
+                                                  collapseClauseOps));
+    }
+
+    rewriter.eraseOp(doLoop);
+    return mlir::success();
+  }
+
+private:
+  /// Collect the list of values used inside the loop but defined outside of it.
+  /// The first item in the retunred list is always the loop's induction
+  /// variable.
+  llvm::SmallVector<mlir::Value>
+  collectLoopLiveIns(fir::DoLoopOp doLoop) const {
```

Comment:
Nit: The general consensus is to avoid hardcoding some arbitrary `llvm::SmallVector` size in the interface of a class (in this case it's the default implicitly-calculated one). I'd generally tend towards doing something like `void collectLoopLiveIns(fir::DoLoopOp doLoop, llvm::SmallVectorImpl<mlir::Value> &liveIns) const`, but I suppose if you really wanted to return the vector and rely on compiler optimizations to avoid copies you could do `  llvm::SmallVectorImpl<mlir::Value> collectLoopLiveIns(fir::DoLoopOp doLoop) const` as well.